fs-storage/src/folder_storage.rs

```diff
@@ -0,0 +1,573 @@
+use serde::{Deserialize, Serialize};
+use std::fs::{self, File};
+use std::io::{Read, Write};
+use std::time::SystemTime;
+use std::{
+    collections::BTreeMap,
+    path::{Path, PathBuf},
+};
+
+use crate::base_storage::{BaseStorage, SyncStatus};
+use crate::monoid::Monoid;
+// use crate::utils::read_version_2_fs;
+use data_error::{ArklibError, Result};
+
+/*
+Note on `FolderStorage` Versioning:
+
+`FolderStorage` is a basic key-value storage system that persists data to disk.
+where the key is the path of the file inside the directory.
+
+
+In version 2, `FolderStorage` stored data in a plaintext format.
+Starting from version 3, data is stored in JSON format.
+
+For backward compatibility, we provide a helper function `read_version_2_fs` to read version 2 format.
+*/
+const STORAGE_VERSION: i32 = 3;
+
+/// Represents a folder storage system that persists data to disk.
+pub struct FolderStorage<K, V>
```

Comment:
In `FileStorage`, how we store the data was straightforward because we know that data is stored as a file with key-value pairs. But for `FolderStorage`, it's not immediately clear how the data is stored. We should add a doc comment to explain it. 

For example, if we're using `bincode` to serialize and deserialize the data where the key is the file name, we should mention that in a comment.